website/blog/2023-04-04-presto-hudi-deep-dive.md

```diff
@@ -0,0 +1,110 @@
+
+<!--truncate-->
+
+## Input formats and Record readers
+Before actually diving into the presto-hudi integration, let us first review a few related concepts. In the Hadoop world, data is stored in a logical file. This file can have various physical layouts (format) on remote storage. A "Path" identifies a file. Further, an "input format" abstraction allows reading portions (splits) of a file. To keep it short and sweet, input formats help generate splits from files which are logical representations of the data. Input formats provide record readers which do the actual reading from these splits.
+
+InputFormat interface exposes [2 methods](https://hadoop.apache.org/docs/stable/api/org/apache/hadoop/mapred/InputFormat.html) - `getSplits()` and `getRecordReader()`.
+
+### Creation of splits
+*`getSplits()`* method facilitates the creation of splits from actual data files. On a high level, this involves reading the input paths from the configuration object and scanning those paths to get the files present. Files are logically represented as FileStatus. These file paths are then passed to the `makeSplit()` method which creates the splits for further processing. Below are the snippets of the actual code where these calls are being made.
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/get-splits-call.png)
+*Figure 1: listStatus() gets called internally from getSplits().*
+
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/make-split-call.png)
+*Figure 2: makeSplit() call is highlighted as part of getSplits() implementation.*
+
+The above snippets are taken from FileInputFormat class. Please note that the lines are highlighted where these calls are being made. 
+
+## Hudi table types
+
+Hudi supports 2 types of tables out of the box, namely, Copy on Write (CoW) and Merge on Read (MoR). CoW is the simpler of the two in terms of the reading and writing complexities involved. Upserts made to this table type result in newer versions of columnar parquet files which can be read using the ParquetPageSource used internally by Presto.
+MoR involves writing incoming upserts to delta log files in avro format by default. These log files are later compacted to create new base parquet files. Since this involves row and columnar format files, this table type supports Read Optimized (RO) and Real Time (RT) queries.
```

Comment:
```suggestion
MoR involves writing incoming updates to delta log files in avro format by default. These log files are later compacted to create new base parquet files. Since this involves row and columnar format files, this table type supports Read Optimized (RO) and Real Time (RT) queries.
```

Reply:
Thank you for pointing this out.